### PR TITLE
Consistent message resolving regarding language fallbacks for all themes

### LINKED ISSUE #15845: Realm localization: Inconsistent message resolving regarding language fallbacks for different themes
### Area

core

### Describe the bug

Currently two different approaches are taken regarding language fallbacks, when realm localization texts are specified.
This is difficult to understand for the users and should be unified.

Furthermore, in contrast to the message resolving of the properties files (see https://github.com/keycloak/keycloak/blob/fc7c57ee12c61730bad089b190b5e2e51c24d3e2/services/src/main/java/org/keycloak/theme/DefaultThemeManager.java#L224), none of these approaches currently considers the fallbacks for variants (e.g. "de-CH-1996" should have fallback "de-CH") and regions (e.g. "de-CH" should have fallback "de").

Currently there are the following approaches.
To keep the text short, I will use the abbreviations **S** (for **S**tatic i18n aka "theme properties") and **RL** (for **R**ealm **L**ocalization).

### Approach 1: "Realm-Localization-Default-Text with lower prio than static i18n"
The prio of messages is the following (highest prio first): **RL \<complete-language-tag>, S \<language>-\<region>-\<variant>, S \<language-region>, S \<language>, S \<default-language>, RL \<default-language>**

This approach is currently used in the following contexts (but doesn't work currently due to https://github.com/keycloak/keycloak/issues/15059):
- login-templates (FreeMarkerLoginFormsProvider)
- email-templates (FreeMarkerEmailTemplateProvider)
- account templates (FreeMarkerAccountProvider)

It is pretty clear that this approach is not very helpful: The purpose of realm localization should be to customize localization, and this is not possible when _RL \<default-language>_ has a lower prio than _S \<default-language>_.
One example: In message_en.properties for our custom login theme, we have a default value of true for key theme.hideLocalizationDropdown. (We decide based on this value whether to hide the localization dropdown.) One of our customers changes this to false, via realm localization text for the default language en. (She wants that all users see the dropdown, regardless of their language.) Then users with language de will still see no dropdown, due to theme.hideLocalizationDropdown still being true, because _S \<default-language>_ has a higher priority than _RL \<default-language>_.

### Current Approach 2: "Realm-Localization-Default-Text with higher prio than static i18n"
The prio of messages is the following (highest prio first): **RL \<complete-language-tag>, RL \<default-language>, S \<language>-\<region>-\<variant>, S \<language-region>, S \<language>, S \<default-language>**

This approach works fine for the example above (theme.hideLocalizationDropdown): It is possible to overwrite the static messages for all languages with the realm localization text of the default locale. But it is questionable whether this is the expected behavior in most cases. The following things should be considered:
- In custom themes, it is possible to overwrite messages from the base theme. For symmetry reasons, it makes sense that realm localization behaves the same as a custom theme, with regards to overwriting message texts. The prio for custom themes is as follows (highest prio first, **SC** = **S**tatic **C**ustom theme, **SB** = **S**tatic **B**ase theme): **SC \<language>-\<region>-\<variant>, SB \<language>-\<region>-\<variant>,  SC \<language-region>, SB \<language-region>, SC \<language>, SB \<language>, SC \<default-language>, SB \<default-language>**
- As a user, I would assume when I specify an English realm localization text for the Account UI title (there might be a typo in the Keycloak default text), this is NOT displayed in the German Account UI page - at least when everything is nicely translated to German. It might be confusing to have a completely German page, with an English title.

### Version

19.0.3

### Expected behavior

The priority of message resolving should be the same in all contexts. A common solution should be implemented that is
- well understood by users
- covers all common scenarios

We suggest to use the same approach that is used when overwriting messages in a sub theme.
The prio of messages would then be (highest prio first): **RL \<language>-\<region>-\<variant>, S \<language>-\<region>-\<variant>, RL \<language-region>, S \<language-region>, RL \<language>, S \<language>, RL \<default-language>, S \<default-language>**

### Actual behavior

Mentioned in description.

### How to Reproduce?

Mentioned in description.

### Anything else?

Solving this issue would also fix https://github.com/keycloak/keycloak/issues/15059.

### PR DESCRIPTION
**Pull request by danielFesenmeyer**
- the prio of messages is now as follows for all themes (RL = realm localization, T = Theme i18n files): RL <variant> > T <variant> > RL <region> > T <region> > RL <language> > T <language> > RL en > T en
- centralize the message resolving logic in helper methods in LocaleUtil and use it for all themes, add unit tests in LocaleUtilTest
- add basic integration tests to check whether realm localization can be used in all supported contexts:
  - Account UI V1: AccountPageTest
  - Account UI V2: org.keycloak.testsuite.ui.account2.InternationalizationTest
  - Admin API messages (currently just used for userprofile validation messages): DeclarativeUserTest
  - Login theme: LoginPageTest
  - Email theme: EmailTest
- deprecate the param useRealmDefaultLocaleFallback=true of endpoint /admin/realms/{realm}/localization/{locale}, because it does not resolve fallbacks as expected and is no longer used in admin-ui v2 (see ui ticket keycloak/keycloak#17679)
- message resolving for admin-ui v2 is implemented in separate ui ticket keycloak/keycloak#17679
- fix locale selection in DefaultLocaleSelectorProvider that a supported region (like "de-CH") will no longer selected instead of a supported language (like "de"), when just the language is requested, add corresponding unit tests

Closes #15845

### CODE REVIEW

**Comment by danielFesenmeyer** [comment]
@mposolda @pedroigor
I have included a commit with the changes for the Admin UI, which were initially implemented in https://github.com/keycloak/keycloak-ui/pull/4395 (for the old UI repo).

I didn't squash the commits yet, cause I thought it's easier to review - just tell me if it's better when I squash them now.

File: server-spi/src/main/java/org/keycloak/theme/Theme.java
Comment (mposolda) [inline server-spi/src/main/java/org/keycloak/theme/Theme.java:78]: How about to instead add to the `Theme` interface some method like `Properties getEnhancedProperties(RealmModel realm, Locale locale)` ?

This can be fine as it looks that all the usages of the method `getGroupedMessages` in this PR are doing something like:
```
Map<Locale, Properties> themeMessages = theme.getGroupedMessages(locale);
Properties messages = LocaleUtil.enhancePropertiesWithRealmLocalizationTexts(realm, locale, themeMessages);
```
which can be simplified IMO just to something like:
```
Properties messages = theme.getEnhancedMessages(realm, locale);
```

In this case, the method can be also declared as `default` on the theme interface with just calling `getProperties(locale)` as that is sufficient for both of `ClassLoaderTheme` and `FolderTheme` IMO and just `ExtendedTheme` will need to implement it in a same way like you're doing?

WDYT?

**Review by mposolda — CHANGES_REQUESTED** [review]
@danielFesenmeyer Thanks a lot for this contribution! And sorry for late review...

I've added few inline comments.

I did not much reviewed admin console related parts @ssilvert Could you please review those or delegate to someone to review?

Question: Do we need some docs for this? I don't know if we need to update localization docs in the "core" guides (like admin console guide and development guide). It seems there is not any docs for the "Realm resource bundles" and it's maybe out of scope of this PR to document this? But at least, I think we may need some short notice in the "Upgrading guide" (possibly for 21.1.0 upgrading guide) as I think that otherwise there will be plenty of bugs reported that localization "resolving" was changed without any notice... WDYT?

**Review by ssilvert — CHANGES_REQUESTED** [review]
@danielFesenmeyer Thanks for the PR.  UI code looks nice except I'd like you to get rid of the `any` type wherever you can.  We try to stay away from `any`.

See https://thoughtbot.com/blog/typescript-stop-using-any-there-s-a-type-for-that

File: services/src/main/java/org/keycloak/services/util/LocaleUtil.java
Comment (danielFesenmeyer) [reply services/src/main/java/org/keycloak/services/util/LocaleUtil.java]: Adjusted as suggested.

File: testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/admin/DeclarativeUserTest.java
Comment (danielFesenmeyer) [reply testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/admin/DeclarativeUserTest.java]: Created a separate ticket with a PR: https://github.com/keycloak/keycloak/issues/19707

File: server-spi/src/main/java/org/keycloak/theme/Theme.java
Comment (danielFesenmeyer) [reply server-spi/src/main/java/org/keycloak/theme/Theme.java]: Thanks for the hint.

Introduced a method Properties `getEnhancedMessages(RealmModel realm, Locale locale)` on the Theme interface and used it instead of `getGroupedMessage`. Makes the usage much simpler.

However, I did not introduce a default implementation, because otherwise the Theme module (keycloak-server-spi) would need a dependency on keycloak-services (due to LocaleUtil). And that dependency would be circular.

**Comment by danielFesenmeyer** [comment]
> @danielFesenmeyer Thanks for the PR. UI code looks nice except I'd like you to get rid of the `any` type wherever you can. We try to stay away from `any`.
> 
> See https://thoughtbot.com/blog/typescript-stop-using-any-there-s-a-type-for-that

@ssilvert Thanks for the reminder. I removed the `any` type whereever possible now - at one place it's still used, but there actually `any` object can be returned (the result of i18next's getResource method).

**Comment by danielFesenmeyer** [comment]
@mposolda 
> Question: Do we need some docs for this? I don't know if we need to update localization docs in the "core" guides (like admin console guide and development guide). It seems there is not any docs for the "Realm resource bundles" and it's maybe out of scope of this PR to document this? But at least, I think we may need some short notice in the "Upgrading guide" (possibly for 21.1.0 upgrading guide) as I think that otherwise there will be plenty of bugs reported that localization "resolving" was changed without any notice... WDYT?

I think it would be good to have some documentation on that topic, but I'm not sure where it should be added.
Let me know when I can assist further.

File: testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/admin/DeclarativeUserTest.java
Comment (mposolda) [reply testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/admin/DeclarativeUserTest.java]: And I see it is closed now. Thanks!

File: server-spi/src/main/java/org/keycloak/theme/Theme.java
Comment (mposolda) [reply server-spi/src/main/java/org/keycloak/theme/Theme.java]: Thanks!

**Review by mposolda — DISMISSED** [review]
@danielFesenmeyer Thanks! I am approving.

I hope to merge this when @ssilvert is ok with the UI changes. Also we need testsuite passing (not 100% sure if test failures in GH actions are related to this PR or random failures...)

**Comment by danielFesenmeyer** [comment]
@mposolda Added change notes. However the CI status check fails, is this due to the flaky cypresss tests? I don't think I changed anything related to those tests.

**Comment by mposolda** [comment]
@ssilvert Do you think this PR is good now? 
@danielFesenmeyer BTV. I see there is still one use of `any` type introduced in this PR in the type `i18n.ts`, which is what @ssilvert preferred to avoid based on some previous comment in this PR? But I am not sure if you guys discussed this already and agreed that this one is good to go?

**Comment by danielFesenmeyer** [comment]
@mposolda, @ssilvert I got rid of the any type wherever possible. There is just one place where it isn't possible from my POV. See my comment above: https://github.com/keycloak/keycloak/pull/17076#issuecomment-1507538229

File: docs/documentation/upgrading/topics/keycloak/changes-22_0_0.adoc
Comment (jonkoops) [inline docs/documentation/upgrading/topics/keycloak/changes-22_0_0.adoc:78]: This is not a Markdown file but AsciiDoc, so I believe you need to use `=` for titles (like in the rest of this file).
```suggestion
= Consistent fallback message resolving for all themes
```

File: js/apps/admin-ui/src/i18n.ts
Comment (jonkoops) [inline js/apps/admin-ui/src/i18n.ts:50]: Perhaps we should be implementing this as a custom backend instead? See the documentation on [Backend Fallback](https://www.i18next.com/how-to/backend-fallback).

**Review by jonkoops — CHANGES_REQUESTED** [review]

**Comment by jonkoops** [comment]
I think this might also be something we need to implement for the Account Console v3. Not sure if we're loading locale overrides at this moment.

File: docs/documentation/upgrading/topics/keycloak/changes-22_0_0.adoc
Comment (danielFesenmeyer) [reply docs/documentation/upgrading/topics/keycloak/changes-22_0_0.adoc]: Thanks for the hint, adjusted that.

**Comment by jonkoops** [comment]
@danielFesenmeyer I see this PR also fixes a bug with the overrides from the message bundles in the Admin Console. I think we should solve this issue separately, so I have started working on a fix for it.

I'll let you know when there is more information available.

File: js/apps/admin-ui/src/i18n.ts
Comment (danielFesenmeyer) [reply js/apps/admin-ui/src/i18n.ts:50]: I also thought of implementing such a fallback, but from my POV it is not possible (at least without a bigger refactoring of i18next):

As far as I understand, the Backend fallback works a bit different as one might expect. It just tries a different (fallback) locale/namespace URL, when the first one cannot be loaded at all (e.g. returns status 404). Hence, it does not work when some message keys are translated and some are not.

**Comment by jonkoops** [comment]
@danielFesenmeyer I've [created](https://github.com/keycloak/keycloak/pull/20183) what I believe might be a proper fix for at least the Admin Console. It will take the namespace into account as you have done here, and expand nested keys that are using the dot separator. Feel free to let me know what you think of this solution in a review, I'd love to get your input on it.

**Comment by jonkoops** [comment]
@danielFesenmeyer could you rebase this PR on `main`? I think you can remove most of the changes of the front-end code for the Admin Console now that #20183 has been merged.

If you are up for it feel free to contribute the updated Cypress tests in a separate PR, as I think they should now work out of the box and that work is more likely to be merged than the changes to the Java code (due to review capacity reasons).

**Review by jonkoops — DISMISSED** [review]
Changes on the UI side LGTM

**Review by mposolda — DISMISSED** [review]
@jonkoops @danielFesenmeyer Thanks for figuring the UI part!
I hope to merge this once tests are ok

**Comment by jonkoops** [comment]
Hmm, looks like not all of the Cypress tests are passing at the moment. Perhaps my implementation is missing something yours did have @danielFesenmeyer?

**Comment by danielFesenmeyer** [comment]
@jonkoops It's failing because there is no translation file for locale "de-CH". It's what about I mentioned here: https://github.com/keycloak/keycloak/pull/20183#discussion_r1188482448.

Could you already talk to the team what they think about supporting locales, for which no translation files exist? Should I create a separate issue for that?

To get this PR merged, I could remove that special test for now.

**Comment by ssilvert** [comment]
> @jonkoops It's failing because there is no translation file for locale "de-CH". It's what I mentioned here: [#20183 (comment)](https://github.com/keycloak/keycloak/pull/20183#discussion_r1188482448).
> 
> Could you already talk to the team what they think about supporting locales, for which no translation files exist? Should I create a separate issue for that?
> 
> To get this PR merged, I could remove that special test for now.

It's fine to comment out the test and create an issue.  It part of our ethos in the UI team.  We don't hold up a PR for one little thing when we can create an issue and move on.

**Comment by jonkoops** [comment]
> Could you already talk to the team what they think about supporting locales, for which no translation files exist? Should I create a separate issue for that?

I think the best option would indeed be to create an issue for that so we can discuss it there. I'd like to avoid blocking this PR.

**Comment by danielFesenmeyer** [comment]
Ok, I've disabled the failing test and created a separate task https://github.com/keycloak/keycloak/issues/20412.

Also rebased once more.

So from my POV everything should be fine now.

**Review by ssilvert — APPROVED** [review]

**Comment by danielFesenmeyer** [comment]
@mposolda @ssilvert @jonkoops thanks for reviewing and merging :)

**Pull request merged**
